Applications/ApplicationsLib/LinearSolverLibrarySetup.h

```diff
@@ -0,0 +1,64 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2015, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#ifndef APPLICATIONSLIB_LINEARSOLVERLIBRARYSETUP_H_
+#define APPLICATIONSLIB_LINEARSOLVERLIBRARYSETUP_H_
+
+namespace ApplicationsLib
+{
+
+/// The LinearSolverLibrarySetup takes care of proper initialization and
+/// shutting down of an external linear solver library. The concrete
+/// implementation is chosen by the build system.
+/// An object of this class must be created at the begining of the scope where
+/// it is used. When the scope closes (or the object is destroyed explicitly)
+/// library shutting down functions are automatically called.
+/// The default implementation is empty providing polymorphic behaviour when
+/// using this class.
+
+#if defined(USE_PETSC)
+#include <petsc.h>
+#include <mpi.h>
+struct LinearSolverLibrarySetup final
+{
+	LinearSolverLibrarySetup(int argc, char* argv[])
+	{
+		MPI_Init(&argc, &argv);
```

Comment:
What is the difference between explicitly calling MPI_Init versus let it be called by PetscInitialize?

From petsc [documentation](http://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Sys/PetscInitialize.html): "PetscInitialize() calls MPI_Init() if that has yet to be called"
And [later](http://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Sys/PetscFinalize.html#PetscFinalize): "MPI_Finalize() is called only if the user had not called MPI_Init() before calling PetscInitialize()."
Se the constructor and destructor are consistent regarding mpi initialization/shutdown.